### PR TITLE
feat(cli): warn of non-existent stacks in `cdk destroy`

### LINKED ISSUE #27179: No exception when stack with wrong cases is deployed
### Describe the bug

When invalid stack name is given to cdk destroy, it does not give an error, but does nothing.

### Expected Behavior

Should give an error: "Stack does not exist"

### Current Behavior

Nothing happens

### Reproduction Steps
Stack "IAMStack" exists, but not "iamstack".

```
> cdk destroy iamstack
Are you sure you want to delete:  (y/n)? y

```

Also notice that it skips the stack name after the word "delete", as it is invalid.

### Possible Solution

Just a check if the stack exists, if not throw error.

### Additional Information/Context

_No response_

### CDK CLI Version

2.89.0 (build 2ad6683)

### Framework Version

_No response_

### Node.js Version

20

### OS

Mac M1 13.5.2 (22G91)

### Language

Typescript

### Language Version

_No response_

### Other information

I'm willing to fix the issue if core team approves :)

### PR DESCRIPTION
**Pull request by go-to-k**
This PR for cli is to warn if stacks with wrong cases (=not exist) specified in `cdk destroy`.

Closes #27179.

----

*By submitting this pull request, I confirm that my contribution is made under the terms of the Apache-2.0 license*

### CODE REVIEW

**Comment by go-to-k** [comment]
Exemption Request: this should be covered with cli-integ tests. In my environment, the behavior was confirmed as expected.

**Comment by vinayak-kukreja** [comment]
Hey, thank you for opening a contribution with us. 
Looks like @SankyRed and @TheRealAmazonKendra were taking a look at the original PR. Will reach out to them if they want to take a look.

**Comment by go-to-k** [comment]
@vinayak-kukreja Thanks. I would be glad for them to see it.

**Comment by go-to-k** [comment]
@lpizzinidev 

Thanks, I changed the message. Please look at this.

**Review by lpizzinidev — APPROVED** [review]
Thanks 👍

**Comment by vinayak-kukreja** [comment]
Hey @go-to-k, apologies that no one has yet got to this. I will take a look today. I have also added a label to not have this PR closed automatically. 
Apologies for the delay.

**Review by vinayak-kukreja — APPROVED** [review]
Hey @go-to-k , the changes looks good to me. I have started a test run for this CLI change in our pipeline. If that succeeds, I will merge this PR in. 

Thank you for this contribution. Really appreciate your patience.

**Comment by go-to-k** [comment]
@vinayak-kukreja 

I changed it to try it out. (But I don't still confirm the behavior because cli-integ-tests itself does not work in my environment.) Please give it a try.

Here are [the tests](https://github.com/aws/aws-cdk/blob/main/packages/%40aws-cdk-testing/cli-integ/tests/cli-integ-tests/cli.integtest.ts#L581-L587).

```ts
['typescript', 'python', 'csharp', 'java'].forEach(language => {
  integTest(`cdk migrate ${language}`, withCDKMigrateFixture(language, async (fixture) => {
    if (language === 'python') {
      await fixture.shell(['pip', 'install', '-r', 'requirements.txt']);
    }

    const stackArn = await fixture.cdkDeploy(fixture.stackNamePrefix, { neverRequireApproval: true, verbose: true, captureStderr: false }, true);
    const response = await fixture.aws.cloudFormation('describeStacks', {
      StackName: stackArn,
    });

    expect(response.Stacks?.[0].StackStatus).toEqual('CREATE_COMPLETE');
    await fixture.cdkDestroy(fixture.stackNamePrefix);
```

In these tests, `true` is specified as the third arg (`skipStackRename`) for `fixture.cdkDeploy`, `fixture.stackNamePrefix` is specified as the first arg (`stackNames`). And, at the `fixture.cdkDestroy`, `fixture.stackNamePrefix` is used.

In `cdkDeploy`, `stackNames` is used instead of `this.fullStackName(stackNames)` when skipStackRename is `true`.

```ts
  public async cdkDeploy(stackNames: string | string[], options: CdkCliOptions = {}, skipStackRename?: boolean) {
    stackNames = typeof stackNames === 'string' ? [stackNames] : stackNames;

    const neverRequireApproval = options.neverRequireApproval ?? true;

    return this.cdk(['deploy',
      ...(neverRequireApproval ? ['--require-approval=never'] : []), // Default to no approval in an unattended test
      ...(options.options ?? []),
      // use events because bar renders bad in tests
      '--progress', 'events',
      ...(skipStackRename ? stackNames : this.fullStackName(stackNames))], options);
  }
```

https://github.com/aws/aws-cdk/blob/main/packages/%40aws-cdk-testing/cli-integ/lib/with-cdk-app.ts#L331

On the other hand, in the `cdkDestroy`, only `this.fullStackName(stackNames)` is used.

```ts
  public async cdkDestroy(stackNames: string | string[], options: CdkCliOptions = {}) {
    stackNames = typeof stackNames === 'string' ? [stackNames] : stackNames;

    return this.cdk(['destroy',
      '-f', // We never want a prompt in an unattended test
      ...(options.options ?? []),
      ...this.fullStackName(stackNames)], options);
  }
```

https://github.com/aws/aws-cdk/blob/main/packages/%40aws-cdk-testing/cli-integ/lib/with-cdk-app.ts#L347

So, in the tests, I changed to pass `fixture.fullStackName(fixture.stackNamePrefix)` instead of `stackNames` into `fixture.cdkDeploy`.

File: packages/@aws-cdk-testing/cli-integ/tests/cli-integ-tests/cli.integtest.ts [hunk 1]
```diff
@@ -578,7 +578,7 @@ integTest('deploy with role', withDefaultFixture(async (fixture) => {
       await fixture.shell(['pip', 'install', '-r', 'requirements.txt']);
     }
 
-    const stackArn = await fixture.cdkDeploy(fixture.stackNamePrefix, { neverRequireApproval: true, verbose: true, captureStderr: false }, true);
+    const stackArn = await fixture.cdkDeploy(fixture.fullStackName(fixture.stackNamePrefix), { neverRequireApproval: true, verbose: true, captureStderr: false }, true);
```
Comment (go-to-k) [inline packages/@aws-cdk-testing/cli-integ/tests/cli-integ-tests/cli.integtest.ts:581]: https://github.com/aws/aws-cdk/pull/27921#issuecomment-1842282221

File: packages/@aws-cdk-testing/cli-integ/tests/cli-integ-tests/cli.integtest.ts [hunk 2]
```diff
@@ -578,7 +578,7 @@ integTest('deploy with role', withDefaultFixture(async (fixture) => {
       await fixture.shell(['pip', 'install', '-r', 'requirements.txt']);
     }
 
-    const stackArn = await fixture.cdkDeploy(fixture.stackNamePrefix, { neverRequireApproval: true, verbose: true, captureStderr: false }, true);
+    const stackArn = await fixture.cdkDeploy(fixture.stackNamePrefix, { neverRequireApproval: true, verbose: true, captureStderr: false });
```
Comment (go-to-k) [inline packages/@aws-cdk-testing/cli-integ/tests/cli-integ-tests/cli.integtest.ts:581]: https://github.com/aws/aws-cdk/pull/27921#issuecomment-1842282221

**Comment by vinayak-kukreja** [comment]
Running in pipeline.

**Comment by vinayak-kukreja** [comment]
I see it failed in pipeline,

```
No stacks match the name(s) cdk-migrate-typescript-integ-0vtozm4x3czs-cdk-migrate-typescript-integ-0vtozm4x3czs
[18:35:22] Error: No stacks match the name(s) cdk-migrate-typescript-integ-0vtozm4x3czs-cdk-migrate-typescript-integ-0vtozm4x3czs
[INTEG TEST::cdk migrate typescript] Failed: Error: 'cdk -v deploy --require-approval=never --progress events cdk-migrate-typescript-integ-0vtozm4x3czs-cdk-migrate-typescript-integ-0vtozm4x3czs' exited with error code 1.

**Comment by go-to-k** [comment]
@vinayak-kukreja 

I see, I think the `cdkDestroy` function should be changed. 

I fixed, please give it a try.

**Comment by vinayak-kukreja** [comment]
Hey it is still failing. I will pull the code and try to take a look today.

**Comment by go-to-k** [comment]
@vinayak-kukreja I see. I agree with you.

If so, what exactly would be the method to log this to the user? For example, there is annotation in construct, but is there an alternative in CLI?

**Comment by go-to-k** [comment]
@vinayak-kukreja 

I've made the change for now, what do you think?

File: packages/aws-cdk/lib/cdk-toolkit.ts [hunk 3]
```diff
@@ -784,7 +787,12 @@ export class CdkToolkit {
       defaultBehavior: DefaultSelection.OnlySingle,
     });
 
-    // No validation
+    const notExistPatterns = selector.patterns.filter(pattern => !stacks.stackArtifacts.find(stack =>
+      minimatch(stack.hierarchicalId, pattern) || (stack.id === pattern && semver.major(versionNumber()) < 2),
+    ));
+    if (notExistPatterns.length > 0) {
+      warning(`${notExistPatterns.join(', ')} ${notExistPatterns.length === 1 ? 'does' : 'do'} not exist.`);
```
Comment (paulhcsun) [inline packages/aws-cdk/lib/cdk-toolkit.ts:794]: The wording in this warning doesn't really tell the user why the stack does not exist. If it's the case that we're just checking for case insensitive matches then maybe we can explicitly tell the user that's the case or point out that they need to change their capitalization. Using the `iamstack` vs `IAMStack` example, something like:

 ```
iamstack does not exist. Do you mean IAMStack?
``` 
or 
```
iamstack does not exist but closely matches IAMStack (or simply another stack name). Please check your casing.
```

**Review by paulhcsun — CHANGES_REQUESTED** [review]

**Comment by go-to-k** [comment]
@paulhcsun 

Thanks for your review. I fixed it and I hope you take a look at it.

And I also changed to not display the message `Are you sure you want to delete:` if there is no matching stack.

destroy-stacks-no-exists

**Comment by paulhcsun** [comment]
@go-to-k Thanks for updating the warning message! It looks a lot better to me. Let me run it by the UX team and I'll let you know if there's any additional feedback they have. I've run the cli-integ tests but one of the regression tests failed. It may just be flaky but I will rerun and update you once that's done.

**Comment by go-to-k** [comment]
@paulhcsun 

> I've run the cli-integ tests but one of the regression tests failed.

Oh...

> And I also changed to not display the message Are you sure you want to delete: if there is no matching stack.

I thought it may be due to [this change](https://github.com/aws/aws-cdk/pull/27921/commits/01d7c2155d5cb383c6d24dca7bcb209c4d571513#diff-8cff125f4b53b025add8740ab74f39ee6c20221c0d3ac5ab7db184a6bb077398R578-R580) that alters the interaction flow, but cli-integ uses [the cdkDestroy](https://github.com/aws/aws-cdk/blob/main/packages/%40aws-cdk-testing/cli-integ/lib/with-cdk-app.ts#L344-L345) with `-f` so it may be not...

Once I see the results of the cli-integ details, I will think about it. Let me know if you discover anything else.

**Comment by go-to-k** [comment]
@paulhcsun 

Could you please tell me the result of the cli integ test? Knowing which regression tests failed, etc., might give us a clue as to how to fix it.
Or are we in a state of waiting for another team?

**Comment by go-to-k** [comment]
@paulhcsun 

Could you please confirm the above comments? Thanks.

**Comment by go-to-k** [comment]
@paulhcsun

Could you please confirm the above comments?

**Comment by paulhcsun** [comment]
Hi @go-to-k, I'm very sorry for not getting back to you on this, we were in a state of waiting for another team but this was on me for not noticing your comments. I'm running the CLI integ tests through the pipeline now and will update you with the results soon. After which I will delegate this review to someone with more knowledge on the CDK CLI.

**Comment by go-to-k** [comment]
@paulhcsun Thank you! I'm looking forward to hearing from you and the team after this!

**Comment by paulhcsun** [comment]
Apologies @go-to-k, there were some additional runs queued in the pipeline before this one so I couldn't check the results in time. It looks like it was skipped for some reason though, I will retry running it now.

**Comment by paulhcsun** [comment]
@go-to-k I've added the `cli-integ-tested` label as the most recent run of the pipeline has passed all checks. Seems like it was just flaky integ tests the last time around. Thanks for your patience.

**Review by paulhcsun — APPROVED** [review]

**Pull request merged**